File: backend/core/src/common/access/project-scoped-crud.service.ts
import { BadRequestException } from '@nestjs/common';
import { In, Repository } from 'typeorm';
import { CrudService } from '../crud/crud.service';
import { ProjectAccessService } from './project-access.service';

export class ProjectScopedCrudService<T extends { id: string; projectId: string }> extends CrudService<T> {
  constructor(
    repo: Repository<T>,
    protected readonly projectAccessService: ProjectAccessService,
  ) {
    super(repo);
  }

  async findAllForUser(userId: string | undefined, projectId?: string): Promise<T[]> {
    if (!userId) return [];
    const projectIds = await this.projectAccessService.filterAccessibleProjectIds(userId, projectId);
    if (projectIds.length === 0) return [];
    return this.repo.find({ where: { projectId: In(projectIds) } as any });
  }

  async findOneForUser(userId: string | undefined, id: string): Promise<T | null> {
    if (!userId) return null;
    const record = await this.findOne(id);
    if (!record) return null;
    await this.projectAccessService.assertProjectAccess(userId, record.projectId);
    return record;
  }

  async createForUser(userId: string | undefined, dto: Partial<T>): Promise<T> {
    if (!userId) {
      throw new BadRequestException('유저 정보가 필요합니다.');
    }
    const projectId = (dto as any).projectId as string | undefined;
    if (!projectId) {
      throw new BadRequestException('projectId가 필요합니다.');
    }
    await this.projectAccessService.assertProjectAccess(userId, projectId);
    return super.create(dto);
  }

  async updateForUser(userId: string | undefined, id: string, dto: Partial<T>): Promise<T | null> {
    if (!userId) return null;
    const record = await this.findOne(id);
    if (!record) return null;
    await this.projectAccessService.assertProjectAccess(userId, record.projectId);
    return super.update(id, dto);
  }

  async removeForUser(userId: string | undefined, id: string): Promise<void> {
    if (!userId) return;
    const record = await this.findOne(id);
    if (!record) return;
    await this.projectAccessService.assertProjectAccess(userId, record.projectId);
    return super.remove(id);
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/common/ProjectScopedCrudService.java
package store.pjcloud.cowrite.core.common;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.data.jpa.domain.Specification;
import store.pjcloud.cowrite.core.service.ProjectAccessService;

public class ProjectScopedCrudService<T extends ProjectScoped> {
    private final ProjectScopedRepository<T> repository;
    private final ProjectAccessService projectAccessService;
    private final ObjectMapper objectMapper;
    private final Class<T> entityClass;

    public ProjectScopedCrudService(ProjectScopedRepository<T> repository,
                                    ProjectAccessService projectAccessService,
                                    ObjectMapper objectMapper,
                                    Class<T> entityClass) {
        this.repository = repository;
        this.projectAccessService = projectAccessService;
        this.objectMapper = objectMapper;
        this.entityClass = entityClass;
    }

    public List<T> findAllForUser(UUID userId, UUID projectId) {
        if (userId == null) return List.of();
        List<UUID> projectIds = projectAccessService.filterAccessibleProjectIds(userId, projectId);
        if (projectIds.isEmpty()) return List.of();
        Specification<T> spec = (root, query, cb) -> root.get("projectId").in(projectIds);
        return repository.findAll(spec);
    }

    public T findOneForUser(UUID userId, UUID id) {
        if (userId == null) return null;
        T record = repository.findById(id).orElse(null);
        if (record == null) return null;
        projectAccessService.assertProjectAccess(userId, record.getProjectId());
        return record;
    }

    public T createForUser(UUID userId, Map<String, Object> body) {
        if (userId == null) throw new IllegalArgumentException("유저 정보가 필요합니다.");
        T entity = objectMapper.convertValue(body, entityClass);
        if (entity.getProjectId() == null) throw new IllegalArgumentException("projectId가 필요합니다.");
        projectAccessService.assertProjectAccess(userId, entity.getProjectId());
        return repository.save(entity);
    }

    public T updateForUser(UUID userId, UUID id, Map<String, Object> body) {
        if (userId == null) return null;
        T existing = repository.findById(id).orElse(null);
        if (existing == null) return null;
        projectAccessService.assertProjectAccess(userId, existing.getProjectId());
        try {
            objectMapper.updateValue(existing, body);
        } catch (com.fasterxml.jackson.databind.JsonMappingException ex) {
            throw new IllegalArgumentException("수정할 값이 올바르지 않습니다.");
        }
        return repository.save(existing);
    }

    public void removeForUser(UUID userId, UUID id) {
        if (userId == null) return;
        T existing = repository.findById(id).orElse(null);
        if (existing == null) return;
        projectAccessService.assertProjectAccess(userId, existing.getProjectId());
        repository.deleteById(id);
    }

    public ProjectScopedRepository<T> repository() {
        return repository;
    }
}
